Added .editorconfig file

filalex77 opened the pull request:
When working on #17, I (again) used spaces instead of tabs as the project tries to.

[EditorConfig](http://EditorConfig.org) is a tool aimed at standardising simple formatting rules for software projects. Here is an `.editorconfig` file that has sensible defaults and uses tabs for completions, C++ source files and `CMakeLists.txt`.

M4444 commented:
Sure, we can add it. I would just reduce it to the essentials:
```
[*]
charset = utf-8
indent_style = tab
tab_width = 8
```
If something more is needed it can always be added later.

Also I'm not sure I understand `insert_final_newline`. Does it just insert an empty line at the end and why would you want that?

filalex77 commented:
It just adds a newline (`\n`) at the end of every file, thus adhering to the POSIX standards. See [Why should text files end with a newline? | StackOverflow](https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline).

filalex77 commented:
Also, I'm not sure on `tab_width = 8`. This is just a visual thing for the editors, and 8 spaces for a tab seems... Overkill. (subjective)

M4444 commented:
> > Also I'm not sure I understand insert_final_newline. Does it just insert an empty line at the end and why would you want that?
> 
> It just adds a newline (`\n`) at the end of every file, thus adhering to the POSIX standards. See [Why should text files end with a newline? | StackOverflow](https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline).

Oh, ok we can add that. I thought it added another empty line.

M4444 commented:
> Also, I'm not sure on `tab_width = 8`. This is just a visual thing for the editors, and 8 spaces for a tab seems... Overkill. (subjective)

It does matter here because after indenting with tabs, spaces are used for precise alignment, and the code is already aligned assuming a tab width of 8. As one example see `src/tmatrix.cpp`, line 56.

Users can always set constant tab width in their editors but it will be their fault if the code isn't aligned for them.

filalex77 commented:
> It does matter here because after indenting with tabs, spaces are used for precise alignment, and the code is already aligned assuming a tab width of 8. As one example see `src/tmatrix.cpp`, line 56.

Thank you for clarifying.

I incorporated your suggestions, and also added `trim_trailing_whitespace = true` - because trailing whitespace never helped anybody.

M4444 commented:
> 
> I incorporated your suggestions, and also added `trim_trailing_whitespace = true` - because trailing whitespace never helped anybody.

I like this in principle, but in markdown files two trailing spaces can (and here in `CONTRIBUTING.md` are) used as a line break. Creating special cases for this would just be messier so I would just leave it out.

Also I'd remove everything above line 8. The style of the config file is pretty verbose by itself so there's no need to explain things further.

filalex77 commented:
Sure, thanks!

M4444 commented:
Nice, thank you!
